### PR TITLE
Rename 'torch/ao/nn/quantized._reference' to 'torch/ao/nn/quantized/reference'.

### LINKED ISSUE #83090: Remove `_` from reference modules path
### 🐛 Describe the bug

Currently the path for reference modules contains `_` which means it's private (https://github.com/pytorch/pytorch/tree/master/torch/nn/quantized/_reference), but we would like to make it public since the reference module is now enabled by default in the fx graph mode quantization flow and it will be added to eager mode flow as well in the future.

Previous:
torch.nn.quantized._reference.Conv2d

After:
torch.nn.quantized.reference.Conv2d

### Versions

master

### PR DESCRIPTION
**Pull request by fufeisi**
Currently, the path for reference modules contains _ which means it's private (https://github.com/pytorch/pytorch/tree/master/torch/ao/nn/quantized/_reference), but we would like to make it public since the reference module is now enabled by default in the fx graph mode quantization flow and it will be added to eager mode flow as well in the future. 

To make '_reference' public, it should satisfy the [public API rules](https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation). 
I did in the first commit (prepare '_reference' to be public):
1: add __all__ to public modules and packages;
2. made functions, that are only used in the file that the function is defined, private by adding _ at their names. 

Fixes #83090. (we rename the 'torch/ao/nn/quantized/_reference', because of migration #81667.)

This is a dup for the #84786.

### CODE REVIEW

**Comment by jerryzh168** [comment]
I feel we can rename torch/nn/quantized/_reference as well just to keep them consistent, why do we skip them?

**Comment by fufeisi** [comment]
> I feel we can rename torch/nn/quantized/_reference as well just to keep them consistent, why do we skip them?

I think we can definitely rename both. Zaf says only rename torch/ao/nn/quantized/_reference because torch/nn/quantized/_reference will be deleted soon.

**Comment by fufeisi** [comment]
I think we should change torch/nn/quantized/_reference as well. Since there is a test (test/quantization/ao_migration/common.py) that checks the consistency of torch/nn and torch/ao/nn.

File: docs/source/quantization-support.rst [hunk 1]
```diff
@@ -590,3 +590,7 @@ the `custom operator mechanism <https://pytorch.org/tutorials/advanced/torch_scr
 .. automodule:: torch.nn.quantizable.modules
 .. automodule:: torch.nn.quantized
    :noindex:
+.. automodule:: torch.ao.nn.quantized.reference
+   :noindex:
+.. automodule:: torch.ao.nn.quantized.reference.modules
```
Comment (jerryzh168) [inline docs/source/quantization-support.rst:595]: I feel we don't want to list this here, it's an implementation detail

File: docs/source/quantization.rst [hunk 2]
```diff
@@ -1156,3 +1156,6 @@ Please take a look at `Limitations of Symbolic Tracing <https://docs-preview.pyt
 .. py:module:: torch.nn.quantized.modules
 .. py:module:: torch.nn.quantized.dynamic
 .. py:module:: torch.nn.quantized.dynamic.modules
+
+.. py:module:: torch.ao.nn.quantized.reference
+.. py:module:: torch.ao.nn.quantized.reference.modules
```
Comment (jerryzh168) [inline docs/source/quantization.rst:1161]: same for this one

File: docs/source/quantization-support.rst (hunk 1 above)
Comment (fufeisi) [reply docs/source/quantization-support.rst:595]: It asks me to either list here or makes it private. My understanding is they need to generate an implement doc for each public module and listing something here is to make a placeholder.

Comment (jerryzh168) [reply docs/source/quantization-support.rst:595]: I see, OK

File: test/quantization/ao_migration/test_ao_migration.py [hunk 3]
```diff
@@ -346,14 +346,14 @@ def test_reference_modules_conv(self):
...
-                                   base='nn.quantized._reference.modules')
+                                   base='nn.quantized.modules')
 
     def test_reference_modules_linear(self):
         function_list = [
             'Linear',
         ]
         self._test_function_import('linear', function_list,
-                                   base='nn.quantized._reference.modules')
+                                   base='nn.quantized.modules')
```
Comment (z-a-f) [inline test/quantization/ao_migration/test_ao_migration.py:356]: This is not the right reference for the test base

File: torch/nn/quantized/reference/modules/conv.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# flake8: noqa: F401
```
Comment (andrewor14) [inline torch/nn/quantized/reference/modules/conv.py:1]: Should we keep these files under `torch/nn/quantized/_reference`? I guess there aren't existing users of this namespace?

**Review by andrewor14 — APPROVED** [review]
Thanks. Looks good!

File: torch/nn/quantized/reference/modules/conv.py (hunk 4 above)
Comment (andrewor14) [reply torch/nn/quantized/reference/modules/conv.py:1]: Never mind, just saw your comment above

**Review by z-a-f — CHANGES_REQUESTED** [review]
Is there a way for us to preserve the git blame history for the `torch/nn/quantized/modules/*`? I understand that `git mv` followed by any edits in the same commit, results in `git rm; git add` sequence. But would be nice to somehow preserve the history.

**Comment by z-a-f** [comment]
> I feel we can rename torch/nn/quantized/_reference as well just to keep them consistent, why do we skip them?

If we are planning on removing the `torch/nn/quantized` version, we can avoid a lot of potentially failing tests by keeping the original location as is, while renaming the new one. In addition to that, `git` doesn't like `git mv` + edits in the file, as it treats it as `git rm` + `git add` new location. That deletes the blame history of that file.

In order to preserve the blame history, I propose we keep the `nn/quantized` location as `_reference`, while updating onlye the `ao/nn/quantized`. @jerryzh168 WDYT?

**Comment by z-a-f** [comment]
You will most likely get `test_public_bindings` fail. There are two ways to fix it:

1. Make sure all files within the `torch/ao/nn/quantized/reference` conform with the [public API rules](https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation)
2. Modify the `allowlist_for_publicAPI.json` to TEMPORARILY exclude the `torch.nn.quantized.reference` from the tests. cc @albanD 

@albanD We could fix the public API directly in here, but that would mean we would have to do `git mv` + modification, which git translates to creating new files. Do you suggest excluding the location in the JSON file or should we create two diffs: one for making it public and another moving it?

**Comment by fufeisi** [comment]
> You will most likely get `test_public_bindings` fail. There are three ways to fix it:
> 
> 1. Make sure all files within the `torch/ao/nn/quantized/reference` conform with the [public API rules](https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation)
> 2. Modify the `allowlist_for_publicAPI.json` to TEMPORARILY exclude the `torch.nn.quantized.reference` from the tests. cc @albanD
> 3. Split into two commits, with the first one preparing the `_reference` to be released to public
> 
> @albanD We could fix the public API directly in here, but that would mean we would have to do `git mv` + modification, which git translates to creating new files. Do you suggest excluding the location in the JSON file or should we create two diffs: one for making it public and another moving it?

Follow option 3 to modify this PR.

**Comment by jerryzh168** [comment]
> If we are planning on removing the `torch/nn/quantized` version, we can avoid a lot of potentially failing tests by keeping the original location as is, while renaming the new one. In addition to that, `git` doesn't like `git mv` + edits in the file, as it treats it as `git rm` + `git add` new location. That deletes the blame history of that file.

blame history for torch/nn/quantized/_reference is not too important since we made the changes in just a few PRs I think, do we plan to delete torch/nn/quantized/_reference soon? if so, I think it doesn't matter too much how we name things under torch/nn/quantized so I'm OK with anything, otherwise I feel it might be better to keep things consistent

**Comment by fufeisi** [comment]
> Is there a way for us to preserve the git blame history for the `torch/nn/quantized/modules/*`? I understand that `git mv` followed by any edits in the same commit, results in `git rm; git add` sequence. But would be nice to somehow preserve the history.

Now we only rename 'torch/ao/nn/quantized/_reference' and avoid any loss of history.

**Review by z-a-f — APPROVED** [review]
Thank you for taking care of the public api section
